### PR TITLE
DEV-1925 If the table fits well within the viewport, border line is n…

### PR DESCRIPTION
**Pull request by KrzysztofZie**
### Context
Fixed a bug where a selection's border edge disappeared when the selection sat flush against a frozen-pane boundary (fixedRowsTop, fixedRowsBottom, fixedColumnsStart). The root cause was that the master  overlay drew that edge exactly on the seam line, where the frozen overlay stacked on top occluded it.

  Selection border edge (selection/border/border.ts)

  1. Added a drawFrozenBoundaryEdge method that detects this case and re-draws the edge inside the appropriate frozen overlay. The edge is split across overlays (top, inline_start, bottom, and both
  corners) and clamped to each one's rendered range, so the segments cover it without overlap or gaps.
  2. Added helper methods drawRowFreezeEdge, drawColumnFreezeEdge, and drawRowFreezeBottomEdge to draw the individual edge slices on the top-row, column, and bottom-row freeze lines.
  3. For a selection corner that lands on both freeze lines at once, drawFrozenBoundaryCorner and drawFrozenBottomBoundaryCorner close the connecting square at the corner.
  4. Added isFrozenBoundaryEdge and isFrozenBottomBoundaryEdge to determine whether a given edge lies on a freeze line and is therefore owned by the frozen overlay.
  5. Added isBoundaryCornerScrolledOut and isBottomBoundaryCornerScrolledOut, which consult the scroll-aware master to stop drawing the edge once the boundary cell scrolls behind the pane.

  Rendering details

  6. A "straddle" approach is used: the master draws the half of the edge below the seam and the overlay draws the half above it, preserving full thickness in both selection and edit modes.
  7. The master hides its seam edge only when the whole selection falls within frozen columns (seamAllFrozenCols), preventing a doubled line.
  8. Correctly handles RTL mode (anchoring via right) and the row/column-0 edge cases, where the line must not protrude past the pane.

  Corner offset (overlay/top.ts)

  9. Added a shouldReserveSelectionCornerOffset method that reserves the selection corner's protruding half-height only when the corner actually falls within the frozen top rows, eliminating a leftover
  edge at the seam (#6937).

  Active header highlight accent (selection/manager.ts, styles/base/_base.scss)

  10. The active-header inline-start accent was drawn with zero width (only reserved on :first-child/:nth-child(2)) and the active row-header accent used a box-shadow that rounded independently and
  doubled on the first row of each overlay. Replaced both with real, layout-aligned gridline borders so every active header shows a matching accent.
  11. Because a header flush with a frozen-pane seam is rendered in a separate overlay table — out of reach of the neighbour :has() rule — SelectionManager now tags the boundary header
  (#markFrozenColumnSeamHeader, #markFrozenTopRowSeamHeader, #markFrozenBottomRowSeamHeader) with -seam / -row-seam-top / -row-seam-bottom classes, which the theme colors to match.

  Tests

  12. Added frozenBoundaryEdge.spec.js with regression coverage across LTR/RTL, all three freeze axes, corner squares, scroll-out, border-thickness, and protrusion edge cases.

### How has this been tested?
npm run test:walkontable -- --testPathPattern=frozenBoundaryEdge

GridSettings add
fixedColumnsLeft: 2,
fixedRowsTop: 2,
fixedRowsBottom: 3

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. https://github.com/handsontable/handsontable/issues/6392
2.
3.

### Affected project(s):
- [x] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [ ] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [ ] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

ClickUp task: https://app.clickup.com/t/9015210959/DEV-1925

---

> [!NOTE]
> **Medium Risk**
> Large, overlay-specific changes to core Walkontable selection border rendering and header styling; behaviour is well covered by new specs but regressions could still show up in frozen-pane + RTL + scroll edge cases.
> 
> **Overview**
> Fixes **missing or doubled selection borders** and **broken active-header highlights** when selections or headers sit on `fixedRowsTop`, `fixedRowsBottom`, or `fixedColumnsStart` seams.
> 
> **Selection borders:** `Border#appear` now delegates seam edges to `drawFrozenBoundaryEdge`, which redraws boundary segments in the owning frozen overlays (top, inline_start, bottom, corners), splits spans across overlays, straddles row/bottom seams between master and overlay where needed, hides duplicate edges elsewhere, drops edges when the boundary cell scrolls out of the master viewport, and closes corner gaps when both freeze lines meet. RTL and thick-border geometry are handled explicitly.
> 
> **Top overlay:** `TopOverlay#shouldReserveSelectionCornerOffset` limits the extra holder height for the fill handle to cases where the focus corner actually lies in frozen top rows (mirroring desktop `cornerVisible`), avoiding a stray line at the seam.
> 
> **Active headers:** `SelectionManager` tags neighbour frozen headers with `-seam` / `-row-seam-top` / `-row-seam-bottom` when the active header is flush with a freeze line; `_base.scss` uses real gridline borders and `:has()` (plus those seam classes) so accents align at overlay boundaries.
> 
> Regression specs: `frozenBoundaryEdge.spec.js`, `frozenBoundarySeamHeader.spec.js`; changelog entry for #12802.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b35129beb0f1f90dd4170250cc040c1b3d6363cc. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by budnix — CHANGES_REQUESTED** [review]
I found some glitches that need to be addressed.

A visual glitch in the corner
Zrzut ekranu 2026-06-22 o 10 30 19

The border persists even when the selection is rendered outside the viewport.
Kapture 2026-06-22 at 10 32 39

**Comment by budnix** [comment]
When `fixedColumnsStart` and `fixedRowsStart` are enabled and the selection is drawn on the top overlay, another glitch occurs.
image

**Comment by budnix** [comment]
When the editor is opened at the edge of the overlay, an additional border is drawn.
Kapture 2026-06-22 at 10 38 18

**Comment by budnix** [comment]
There is border misalignment on the the left when autofill feature is used
Kapture 2026-06-22 at 10 40 18

File: handsontable/src/3rdparty/walkontable/src/selection/border/border.ts [hunk 1]
```diff
@@ -658,6 +1015,29 @@ class Border {
...
+    // cell in the first non-frozen column but a frozen row), so both must hide it.
+    const overlayName = wtTable.name;
+
+    if (this.isFrozenBoundaryEdge('row', originalFromRow) &&
+        (wtTable.isMaster || overlayName === 'inline_start')) {
+      this.topStyle!.display = 'none';
+    }
+    if (this.isFrozenBoundaryEdge('column', originalFromColumn) &&
+        (wtTable.isMaster || overlayName === 'top')) {
+      this.startStyle!.display = 'none';
```
[the remark this answers is not in the document]
Comment (KrzysztofZie) [reply handsontable/src/3rdparty/walkontable/src/selection/border/border.ts:1175]: Not a bug. The freeze-edge is always pinned to the freeze line, and isBoundaryCornerScrolledOut is true exactly when the boundary cell has scrolled behind the frozen pane — so its edge sits behind the
  pane and drawing nothing on the seam is the correct result. The master's hide must stay unconditional: when the boundary scrolls out, appear clamps fromRow/fromColumn and would otherwise paint a
  spurious edge at the top/start of the rendered range, so gating the hide on scroll state would actually introduce that stray line, not fix a gap.

File: handsontable/src/3rdparty/walkontable/src/selection/border/border.ts [hunk 2]
```diff
@@ -499,6 +499,352 @@ class Border {
...
+        return true;
+      }
+    }
+
+    // The column-freeze edge (selection inline-start edge on the column freeze line) is owned by the
+    // `inline_start` overlay across the non-frozen rows and by the corner overlay across the frozen
+    // rows the selection reaches. Mirror of the branch above on the column axis.
+    if (columnEdgeOwned && (isInlineStartOverlay || isCornerOverlay)) {
+      const firstRow = Math.max(fromRow, wtTable.getFirstRenderedRow());
+      const lastRow = Math.min(toRow, wtTable.getLastRenderedRow());
```
[the remark this answers is not in the document]
Comment (KrzysztofZie) [reply handsontable/src/3rdparty/walkontable/src/selection/border/border.ts:651]: Not a bug — the finding is based on a stale premise. The row-freeze branch no longer uses getFirstVisibleColumn; both branches now clamp with the rendered range
  (getFirstRenderedColumn/getFirstRenderedRow), so they are symmetric and the described asymmetry doesn't exist. Rendered is intentional here: the visible bound drops the boundary cell on a partial
  scroll (first > last → the edge disappears instead of staying pinned to the freeze line), while occlusion behind the pane is handled separately by columnEdgeOwned via isBoundaryCornerScrolledOut.

**Comment by qunabu** [comment]
Task linked: [DEV-1925 If the table fits well within the viewport, border line is not drawn on an edge of an overlay](https://app.clickup.com/t/9015210959/DEV-1925)

**Comment by budnix** [comment]
@KrzysztofZie it's much better. Two things I'd try to fix. First one I'd propose fixing the border for focused cell for top overlay:
Kapture 2026-06-24 at 09 21 50

and the second the fix also should be applied to the bottom overlay:

Kapture 2026-06-24 at 09 23 39

**Review by budnix — CHANGES_REQUESTED** [review]
https://github.com/handsontable/handsontable/pull/12802#issuecomment-4786861338

**Comment by budnix** [comment]
Also found that there is misalignment on the edge of the left overlay when a whole column is selected and there is missing borders for left-bottom corner:
Zrzut ekranu 2026-06-25 o 08 16 58

Same for rows:
image

File: handsontable/src/3rdparty/walkontable/src/overlay/top.ts [hunk 3]
```diff
@@ -225,6 +226,51 @@ export class TopOverlay extends Overlay {
...
+   * enabled (`cornerVisible` truthy), mirroring `Border#appear`.
+   *
+   * @private
+   * @param {object|null} focusSelection The focus Selection instance (or `null` when none).
+   * @returns {boolean}
+   */
+  shouldReserveSelectionCornerOffset(
+    focusSelection: {
+      settings: Record<string, unknown>;
+      cellRange?: { getBottomEndCorner: () => { row: number | null } } | null;
```
Comment (budnix) [inline handsontable/src/3rdparty/walkontable/src/overlay/top.ts]: The `cellRange` field here is a hand-written structural type that mirrors only part of `CellRange`. A future rename of `getBottomEndCorner` on the real class will not be caught here — TypeScript keeps compiling the old shape and the call silently breaks at runtime. The cleanest fix is to collapse the whole param to `Selection | null` (import type from `../selection/selection`), which gives `CellRange | null` for `cellRange` and `CellCoords` for the return of `getBottomEndCorner()` for free. At minimum the field should use `CellRange | null` (import from `../cell/range`).

```suggestion
      cellRange?: CellRange | null;
```

File: handsontable/src/3rdparty/walkontable/src/overlay/top.ts [hunk 4]
```diff
@@ -225,6 +226,51 @@ export class TopOverlay extends Overlay {
...
+    if (bottomEndRow === null || bottomEndRow >= fixedRowsTop) {
+      return false;
+    }
+
+    if (isMobileBrowser()) {
+      return true;
+    }
+
+    const border = focusSelection.settings.border as
+      { cornerVisible?: boolean | ((...args: unknown[]) => boolean) } | undefined;
```
Comment (budnix) [inline handsontable/src/3rdparty/walkontable/src/overlay/top.ts]: The inline cast type is already exported verbatim as `BorderInstanceSettings["border"]` in `selection/border/types.ts` — same `cornerVisible` shape, same optional fields. If that interface ever gains a second overload for `cornerVisible`, this cast stays on the old definition and `!!cornerVisible(layerLevel)` would call it with the wrong contract, quietly reversing the fix. Replace with `BorderInstanceSettings["border"]` (import type from `../selection/border/types`).

```suggestion
      BorderInstanceSettings["border"];
```

**Review by budnix — COMMENTED** [review]
Curious whether `inlineStart.ts` needs the same guard. The symmetric method `adjustRootChildrenSize` in `inlineStart.ts:222` still unconditionally reserves `parseInt(cornerStyle.width, 10) / 2` whenever any focus selection exists — it did not get a `shouldReserveSelectionCornerOffset`-style predicate like `top.ts` did here. On a grid where `fixedColumnsStart` is set and the table fits within the viewport horizontally, that could leave a leftover vertical line at the inline-start overlay edge — same class of rendering defect, different axis.

**Review by budnix — APPROVED** [review]
